src/solvers/mg_solver.hpp

```diff
@@ -76,13 +77,18 @@ class MGSolver {
   PARTHENON_INTERNALSOLVERVARIABLE(u, temp); // Temporary storage
   PARTHENON_INTERNALSOLVERVARIABLE(u, u0);   // Storage for initial solution during FAS
   PARTHENON_INTERNALSOLVERVARIABLE(u, D);    // Storage for (approximate) diagonal
+
+  using internal_types_tl = TypeList<res_err, temp, u0, D>;
   std::vector<std::string> GetInternalVariableNames() const {
-    return {res_err::name(), temp::name(), u0::name(), D::name()};
+    std::vector<std::string> names;
+    internal_types_tl::IterateTypes(
+        [&names](auto t) { names.push_back(decltype(t)::name()); });
+    return names;
   }
 
   MGSolver(StateDescriptor *pkg, MGParams params_in, equations eq_in = equations(),
-           std::vector<int> shape = {})
-      : params_(params_in), iter_counter(0), eqs_(eq_in) {
+           std::vector<int> shape = {}, const std::string &container = "base")
+      : params_(params_in), iter_counter(0), eqs_(eq_in), container_(container) {
```

Comment:
Just thinking out loud here (independent of this PR): We are still sharing comm buffer of each variable across containers, aren't we?
Do we have any control over ensuring that only one communication piece is running/using the comm buffers in a task list at any given time?

Reply:
We don't have any control over that and I agree that if you had a task list that had independent groups of tasks that worked on different containers but also communicated you could possibly run into issues where one container grabs messages meant for another container. That is similar to some of the issues I had to fix for multi-grid in this PR. I am not sure what the best approach to dealing with this is. Making buffers for each container doesn't seem ideal and it could be a little tricky to build buffers and MPI tag maps for multiple containers in the current framework (since now buffers are built right after remesh, not at the point when containers are first asked for). Maybe we should put up a warning issue about this.